src/lib/Microsoft.Health.Fhir.Ingest.Template/CalculatedFunction/CalculatedFunctionContentTemplate.cs

```diff
@@ -0,0 +1,210 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using EnsureThat;
+using Microsoft.Health.Fhir.Ingest.Data;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Ingest.Template.CalculatedFunction
+{
+    public class CalculatedFunctionContentTemplate : IContentTemplate
+    {
+        public const string MatchedToken = "matchedToken";
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string TypeName { get; set; }
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string TypeMatchExpression { get; set; }
+
+        public virtual ExpressionLanguage? TypeMatchExpressionLanguage { get; set; }
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string DeviceIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? DeviceIdExpressionLanguage { get; set; }
+
+        public virtual string PatientIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? PatientIdExpressionLanguage { get; set; }
+
+        public virtual string EncounterIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? EncounterIdExpressionLanguage { get; set; }
+
+        [JsonProperty(Required = Required.Always)]
+        public virtual string TimestampExpression { get; set; }
+
+        public virtual ExpressionLanguage? TimestampExpressionLanguage { get; set; }
+
+        public virtual string CorrelationIdExpression { get; set; }
+
+        public virtual ExpressionLanguage? CorrelationIdExpressionLanguage { get; set; }
+
+#pragma warning disable CA2227
+        public virtual IList<CalculatedFunctionValueExpression> Values { get; set; }
+#pragma warning restore CA2227
+
+        [JsonConverter(typeof(StringEnumConverter))]
+        public virtual ExpressionLanguage DefaultExpressionLanguage { get; set; } = ExpressionLanguage.JsonPath;
+
+        [JsonIgnore]
+        public virtual IExpressionEvaluatorFactory ExpressionEvaluatorFactory { get; set; } = new CachingExpressionEvaluatorFactory();
+
+        public virtual IEnumerable<Measurement> GetMeasurements(JToken token)
+        {
+            EnsureArg.IsNotNull(token, nameof(token));
+            EnsureArg.IsTrue(token is JObject);
+
+            foreach (var typeToken in MatchTypeTokens(token as JObject))
+            {
+                yield return CreateMeasurementFromToken(typeToken);
+            }
+        }
+
+        protected T EvalExpression<T>(JToken token, params Expression[] expressions)
```

Comment:
It seems like a lot of this code is duplicated from JsonPathContentTemplate; I'm wondering if it makes sense to either move the duplicate code to another class, or have the 2 inherit from a common class.

Reply:
I chose to leave them separate for now, as this new content template essentially replaces `JsonPathContentTemplate`. I could update it to make use of the new `Expression` class and then colocate common features, but since its legacy it feels like overkill. I could be persuaded otherwise, though.

Reply:
After thinking about it a bit more the JsonPathContentTemplate and its subclass could benefit from the additional logging that I've added into the new `CalculatedContentTemplate` class. This could prove useful to folks who don't migrate right away. 
I'll go ahead with the refactor. If its the only thing holding up this PR I'll do it within another PR.

Reply:
I've refactored the code to separate out the data from the behavior for ContentTemplates. Behavior has been moved into a class called `MeasurementExtractor` and the ContentTemplates themselves are just simple data containers.